### DISCUSSION TITLE
如何继承`element-plus`组件的`props`和`emits`类型?

### DISCUSSION THREAD

**Discussion by ckvv** [element-plus/element-plus#8394] [Q&A] [answered]
如下我封装了一个带有label的input组件

```vue
<script setup lang="ts">
import { ElInput } from "element-plus";
import type { InputEmits, InputProps } from "element-plus";

defineProps<
  InputProps & {
    label?: string;
  }
>();
defineEmits<InputEmits>();
</script>

<template>
  <div class="w-input">
    {{ label }}
    <el-input v-bind="$attrs">
      <template #[slotName]="slotProps" v-for="(slot, slotName) in $slots">
        <slot :name="slotName" v-bind="slotProps" />
      </template>
    </el-input>
  </div>
</template>
```

在引用时抛出了以下异常
```
Type '{}' is not assignable to type 'IntrinsicAttributes & Partial<{}> & Omit<Readonly<ExtractPropTypes<__VLS_TypePropsToRuntimeProps<{ readonly type: string; readonly disabled: boolean; readonly modelValue: EpPropMergeType<(new (...args: any[]) => (string | number | null | undefined) & {}) | (() => string | ... 2 more ... | undefined) | ((new (...args...'.
  Type '{}' is missing the following properties from type 'Omit<Readonly<ExtractPropTypes<__VLS_TypePropsToRuntimeProps<{ readonly type: string; readonly disabled: boolean; readonly modelValue: EpPropMergeType<(new (...args: any[]) => (string | number | null | undefined) & {}) | (() => string | number | null | undefined) | ((new (...args: any[]) => (string | ... 2 more ... ...': type, disabled, modelValue, autosize, and 11 more.
```

**Comment by ckvv** [comment]
似乎vue目前不支持这么做！！！
https://github.com/vuejs/core/issues/4294

**Comment by ckvv** [comment]
改成`Options API`可以了，不知道有没有更好的方法？
```vue
<script lang="ts">
import { defineComponent } from "vue";
import { inputProps, inputEmits } from "element-plus/lib/";

export default defineComponent({
  props: {
    ...inputProps,
    label: String,
  },
  emits: inputEmits,
});
</script>

<template>
  <div class="w-input">
    {{ label }}
    <el-input v-bind="$attrs">
      <template #[slotName]="slotProps" v-for="(slot, slotName) in $slots">
        <slot :name="slotName" v-bind="slotProps" />
      </template>
    </el-input>
  </div>
</template>
```

**Comment by sxzz** [comment, marked as answer]
```vue
<script setup lang="ts">
import { ElInput, inputProps, inputEmits } from "element-plus";

defineProps({
    ...inputProps,
    label: String,
});
defineEmits(inputEmits);
</script>
```

**Reply by baixiaoyu2997** [reply]
假如我们称呼以上第一层封装组件为A。
当我要再次封装A写一个B的时候，我没办法在B中继续使用`...inputProps`的方式，如果我在A设置了一个带有默认值的prop，这样会覆盖掉那个prop。
或许为每一个组件单独写一个prop文件是一个比较好的方法？

**Reply by sxzz** [reply]
export 出 A 的 props，SFC 不支持 export，所以要在一个普通的 TS 文件中 export。

**Comment by yangmiao7** [comment]
暴露的Expose有什么方法直接继承吗？我现在都是写函数再包装一层。

**Reply by ckvv** [reply]
可以暴露内部的ref

**Reply by huangbh1024** [reply]
![image](https://github.com/element-plus/element-plus/assets/72372968/e078df80-afc0-4e3b-8339-01deccd07bc1)
这样子可以暴露expose

**Reply by wencaizhang** [reply]
> ![image](https://private-user-images.githubusercontent.com/72372968/285078949-e078df80-afc0-4e3b-8339-01deccd07bc1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM1MzQxNzYsIm5iZiI6MTczMzUzMzg3NiwicGF0aCI6Ii83MjM3Mjk2OC8yODUwNzg5NDktZTA3OGRmODAtYWZjMC00ZTNiLTgzMzktMDFkZWNjZDA3YmMxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjA3VDAxMTExNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkwMmMyMWFjZWNhOTAxMGM0ZmFiMWFkNTA5YWNkZGI4YjRiMTQzMzRmZmU5NGQ0ZjJhMTVjNjdiNzk1MzU5ZTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Npxnc9wIRR9oY-xZhs3JacbLzagMxRNZVtKKl3z_QF8) 这样子可以暴露expose

@huangbh1024  请问为什么要写成 `get(target, key) ?? get(formRef.value || {}, key)` 而不是 `get(formRef.value || {}, key)`。

我的理解是这里的目的是想把 `formRef.value` 的属性暴露出去，而`target` 我们知道已经是 `{}` ，为什么优先从 `target` 上面获取属性 `key` 呢？

**Reply by huangbh1024** [reply]
因为target内可以添加一些自己需要暴露的方法
